File: scala/sdktest/OutputListOfTransactionsHistoryEntityTest.scala
// Generated basic-flow test for the output_list_of_transactions_history entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped OutputListOfTransactionsHistoryTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.bluefintecsuserbackofficesdk.core.{Helpers, BluefinTecsUserBackofficeSDK}
import voxgig.bluefintecsuserbackofficesdk.utility.struct.Struct

object OutputListOfTransactionsHistoryEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("output_list_of_transactions_history.instance") {
      val testsdk = BluefinTecsUserBackofficeSDK.testSDK()
      val ent = testsdk.outputListOfTransactionsHistory(null)
      rep.check("output_list_of_transactions_history.instance", ent != null, "expected non-null output_list_of_transactions_history entity")
    }

    rep.scope("output_list_of_transactions_history.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/output_list_of_transactions_history/OutputListOfTransactionsHistoryTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = BluefinTecsUserBackofficeSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("output_list_of_transactions_history01", "OUTPUT_LIST_OF_TRANSACTIONS_HISTORY01")
      idmap.put("output_list_of_transactions_history02", "OUTPUT_LIST_OF_TRANSACTIONS_HISTORY02")
      idmap.put("output_list_of_transactions_history03", "OUTPUT_LIST_OF_TRANSACTIONS_HISTORY03")
      val now = System.currentTimeMillis()

      // CREATE
      val outputListOfTransactionsHistoryRef01Ent = client.outputListOfTransactionsHistory(null)
      var outputListOfTransactionsHistoryRef01Data = Helpers.toMapAny(Struct.getprop(
          Struct.getpath(entityData, "new.output_list_of_transactions_history"), "output_list_of_transactions_history_ref01"))
      val outputListOfTransactionsHistoryRef01DataResult = outputListOfTransactionsHistoryRef01Ent.create(outputListOfTransactionsHistoryRef01Data, null)
      outputListOfTransactionsHistoryRef01Data = Helpers.toMapAny(outputListOfTransactionsHistoryRef01DataResult)
      rep.check("output_list_of_transactions_history.create.map", outputListOfTransactionsHistoryRef01Data != null, "expected create result to be a map")
    }
  }
}

File: cpp/test/output_list_of_transactions_history_entity_test.cpp
// Generated basic-flow test for the output_list_of_transactions_history entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct OutputListOfTransactionsHistorySetup {
  std::shared_ptr<BluefinTecsUserBackofficeSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static OutputListOfTransactionsHistorySetup output_list_of_transactions_history_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/output_list_of_transactions_history/OutputListOfTransactionsHistoryTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = BluefinTecsUserBackofficeSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("output_list_of_transactions_history01"), Value("output_list_of_transactions_history02"), Value("output_list_of_transactions_history03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"BLUEFIN_TECS_USER_BACKOFFICE_TEST_OUTPUT_LIST_OF_TRANSACTIONS_HISTORY_ENTID", idmap},
    {"BLUEFIN_TECS_USER_BACKOFFICE_TEST_LIVE", Value("FALSE")},
    {"BLUEFIN_TECS_USER_BACKOFFICE_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "BLUEFIN_TECS_USER_BACKOFFICE_TEST_OUTPUT_LIST_OF_TRANSACTIONS_HISTORY_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  bool live = getp(env, "BLUEFIN_TECS_USER_BACKOFFICE_TEST_LIVE") == Value("TRUE");

  OutputListOfTransactionsHistorySetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void output_list_of_transactions_history_entity_instance() {
  auto testsdk = BluefinTecsUserBackofficeSDK::testSDK();
  auto ent = testsdk->output_list_of_transactions_history();
  ASSERT_EQ(ent->getName(), std::string("output_list_of_transactions_history"), "entity name");
}


static void output_list_of_transactions_history_entity_basic() {
  auto setup = output_list_of_transactions_history_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"create"}) {
    auto sk = is_control_skipped("entityOp", std::string("output_list_of_transactions_history.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;
  // CREATE
  auto output_list_of_transactions_history_ref01_ent = client->output_list_of_transactions_history();
  Value output_list_of_transactions_history_ref01_data = Helpers::toMapAny(getp(Struct::getpath(setup.data, {"new", "output_list_of_transactions_history"}), "output_list_of_transactions_history_ref01"));
  if (!output_list_of_transactions_history_ref01_data.is_map()) output_list_of_transactions_history_ref01_data = vmap();
  {
    Value output_list_of_transactions_history_ref01_data_result = output_list_of_transactions_history_ref01_ent->create(Struct::clone(output_list_of_transactions_history_ref01_data), Value::undef())->data();
    output_list_of_transactions_history_ref01_data = Helpers::toMapAny(output_list_of_transactions_history_ref01_data_result);
    if (!output_list_of_transactions_history_ref01_data.is_map()) output_list_of_transactions_history_ref01_data = vmap();
    ASSERT_TRUE(output_list_of_transactions_history_ref01_data.is_map(), "expected create result to be a map");
  }

}

int main() {
  T_RUN(output_list_of_transactions_history_entity_instance);
  T_RUN(output_list_of_transactions_history_entity_basic);
  return sdktest::summary("output_list_of_transactions_history_entity_test");
}
